ludwig/schema/llms/base_model.py

```diff
@@ -0,0 +1,166 @@
+from abc import ABC
+from typing import Optional, Type
+
+from ludwig.api_annotations import DeveloperAPI
+from ludwig.error import ConfigValidationError
+from ludwig.schema import utils as schema_utils
+from ludwig.schema.metadata import LLM_METADATA
+from ludwig.schema.metadata.parameter_metadata import ParameterMetadata
+from ludwig.schema.utils import ludwig_dataclass
+from ludwig.utils.registry import Registry
+
+MODEL_PRESETS = {
+    "opt-350m": "facebook/opt-350m",
+    "opt-1.3b": "facebook/opt-1.3b",
+    "pythia-2.8b": "EleutherAI/pythia-2.8b",
+    "gpt-neo-2.7B": "EleutherAI/gpt-neo-2.7B",
+    "bloomz-3b": "bigscience/bloomz-3b",
+    "gpt-j-6b": "EleutherAI/gpt-j-6b",
+    "stablelm-base-alpha-3b": "stabilityai/stablelm-base-alpha-3b",
+    "llama-7b": "huggyllama/llama-7b",
+    "vicuna-7b": "eachadea/vicuna-7b-1.1",
+    "bloom-7b": "bigscience/bloom-7b1",
+    "stablelm-base-alpha-7b": "stabilityai/stablelm-base-alpha-7b",
+    "pythia-12b": "EleutherAI/pythia-12b",
+    "oasst-sft-1-pythia-12b": "OpenAssistant/oasst-sft-1-pythia-12b",
+    "vicuna-13b": "eachadea/vicuna-13b-1.1",
+}
+
+
+base_model_registry = Registry()
+
+
+@DeveloperAPI
+def register_base_model(name: str):
+    def wrap(config: BaseModelConfig):
+        base_model_registry[name] = config
+        return config
+
+    return wrap
+
+
+@DeveloperAPI
+@ludwig_dataclass
+class BaseModelConfig(schema_utils.BaseMarshmallowConfig, ABC):
+    preset: Optional[str]
+    name: str
+
+
+@DeveloperAPI
+@register_base_model(name=None)
+@register_base_model(name="none")
+@ludwig_dataclass
+class NoPresetModelConfig(BaseModelConfig):
+    def __post_init__(self):
+        if not self.name:
+            raise ConfigValidationError(
+                "LLM requires `base_model.name` to be set. This can be any pretrained CausalLM on huggingface. "
+                "See: https://huggingface.co/models?pipeline_tag=text-generation&sort=downloads"
+            )
+
+    preset: Optional[str] = schema_utils.StringOptions(
+        options=["none"],
+        default=None,
+        allow_none=True,
+        parameter_metadata=LLM_METADATA["base_model"]["preset"],
+    )
+
+    name: str = schema_utils.String(
+        default=None,
+        allow_none=True,
+        description=(
+            "The name of the model to use. This can be a local path or a "
+            "remote path. If it is a remote path, it must be a valid HuggingFace "
+            "model name. If it is a local path, it must be a valid HuggingFace "
+            "model name or a path to a local directory containing a valid "
+            "HuggingFace model."
```

Comment:
Should we include a longer description with a link to the HF models?

Reply:
We can do this, yeah. I would do it in the parameter metadata, which is consistent with how we do it for text encoders, etc. We can do it in a follow-up.